### ISSUE TITLE
Export the server constants to client.

### ISSUE THREAD

**Issue by yahu1031** [serverpod/serverpod#708]
## Is your feature request related to a problem? Please describe. ##
Exporting serverside constants/functions(which are not being generated in client code) to client code.

```dart
// Server side - my_endpoint_file.dart
Future<bool> generateQrCode(Session session, String qrData, QrBodyType qrbodyType) async {
  // my function
}

/// QR Code body type - *_server/lib/src/consts/enums.dart
enum QrBodyType {
  square,
  ...
}
```

```dart
// client

import 'package:fluttermeetups_client/src/consts/enum.dart' as _iy; // This doesn't exist.
...

_ix.Future<bool> generateQrCode(
  String qrData,
  _iy.QrBodyType qrbodyType,
) =>
    caller.callServerEndpoint<bool>(
      'qrEndPoint',
      'generateQrCode',
      {
        'qrData': qrData,
        'qrbodyType': qrbodyType,
      },
    );
```

```dart
// When I try accessing the function from the flutter app
// I see it as dynamic cause the import doesn't work
Future<bool> generateQrCode(String qrData, dynamic qrbodyType)
```

--- 

> ## NOTE: ##
> This issue/enhancement req is not for enums. It is for custom datatypes in the server must be accessed by the client & app also.

**Comment by fischerscode** [comment]
I've always created a shared package used by the server and the client for something like this. But in my approach, one can not use classes generated by Serverpod.

We could introduce an annotation that indicates that a constant, function or type should be copied to the client code. But I think there are edge cases where it's hard to use the right imports.

**Comment by yahu1031** [comment]
True. But the problem with having a shared package is like, we need to add one more package(shared) to the pubspec. I don't wanna add something for just constants. If this can go through the server and client(generated variables from the server), that would be ease. Like we. do the classes. Such that we can also utilize the exception types from the server without writing things as models(My approach).

**Comment by vlidholt** [comment]
Maybe there should be a way to declare constants in the yaml files also in the future, to also be able to provide default values. It's not super easy to implement and also to keep the syntax clean.

For now, a sufficient solution as @fischerscode mentions is to have a shared package between client and server. It has the benefit that you can put any shared code there too.

**Comment by blopker** [comment]
Is it reasonable to put the shared types in the client package and have the server import that? I'm also interested in a place for shared logic, like form validation or state transitions.

**Comment by fischerscode** [comment]
It should work, but it's definitely not the most elegant solution. I'd really recommend using a new package.

Shared logic can be placed in any common package.

**Comment by blopker** [comment]
I'm new to Severpod, but already see a lot of great things about it. However, there are for sure a lot of downsides to using Dart on the server. Like lack of observability tooling, libraries and first class platform support (AWS, Google Cloud, etc.) for example. 

On the other hand, to me the main reason to adopt Serverpod, and Dart on the server in general, is code sharing with Flutter. Our backend is currently written in Firebase Functions and there's so much boilerplate to bridge JS and Dart, not to mention unneeded server calls to do validation. I think already Serverpod does a great job with (de)serialization across platforms, which is probably the number 1 pain point we have. Although, I would love to keep virtually all our business logic in a shared space as well. Having that supported in Serverpod as a main feature would be a huge plus for us.

I don't totally understand why putting shared logic in the client package is problematic, other than the semantics around the naming, but I wanted to share my experience so far in case it's at all helpful! 😀

**Comment by fischerscode** [comment]
> I don't totally understand why putting shared logic in the client package is problematic

It's just not a best practice since you would have code that is not needed on the server (client - server communication, client side generated classes) on the server. In the best case, the compiler gets rid of these parts. In the worst case, these parts bloat you server-side binary and require additional resources. Additionally, you'll always have to make sure you're using the correct generated classes (the names are the same on the client and server side).

Also, there is almost no cost in using a separate package.  Especially since every Serverpod project is already a multi package project.

**Comment by blopker** [comment]
I understand the bloat issue in general, but in this case, the `serverpod` package (what the app server depends on) already depends on `serverpod_client`, which is the only package the app's client package depends on. So the only additional bloat is from the generated types in the app client package. However, that code isn't bloat since the shared code I'd want to write would need access to those types.

I took a look at the generated code and there does seem to be some differences in the types generated for the server vs client. It seems to me the main differences are when types are used for database access on the server, where the types extend `TableRow` and have additional methods for the database.

Is there a way we can use the same type definitions on the server and client, so we can write shared logic that works with the generated types on both sides?
